source/byte_buf.c

```diff
@@ -0,0 +1,109 @@
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/byte_buf.h>
+#include <assert.h>
+
+int aws_string_split_on_char(struct aws_byte_buf *input_str, char split_on, struct aws_array_list *output) {
+    assert(input_str);
+    assert(output);
+    assert(output->item_size >= sizeof(struct aws_byte_buf));
+
+    size_t current_pos = 0;
+    size_t last_offset = 0;
+
+    for(; current_pos < input_str->len; ++current_pos) {
+        if(current_pos < input_str->len - 2 && input_str->buffer[current_pos] == (uint8_t)split_on) {
+            struct aws_byte_buf buffer = {
+                    .buffer = input_str->buffer + last_offset,
+                    .len = current_pos - last_offset,
+            };
+
+            if (AWS_UNLIKELY(aws_array_list_push_back(output, (const void *)&buffer))) {
+                return AWS_OP_ERR;
+            }
+
+            last_offset = current_pos + 1;
+        }
+    }
+
+    if (last_offset < input_str->len) {
+        struct aws_byte_buf buffer = {
+                .buffer = input_str->buffer + last_offset,
+                .len = current_pos - last_offset,
+        };
+
+        if (AWS_UNLIKELY(aws_array_list_push_back(output, (const void *)&buffer))) {
+            return AWS_OP_ERR;
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_string_split_on_str(struct aws_byte_buf *input_str, char *split_on,
+                            struct aws_array_list *output) {
+    assert(input_str);
+    assert(output);
+    assert(output->item_size >= sizeof(struct aws_byte_buf));
+    assert(split_on);
+    size_t token_len = strlen(split_on);
+    assert(token_len);
+
+    size_t current_pos = 0;
+    size_t last_offset = 0;
+
+    int8_t possible_match_idx = 0;
+
+    for(; current_pos < input_str->len; ++current_pos) {
+
+        if (current_pos < input_str->len - 2 && input_str->buffer[current_pos] == (uint8_t)split_on[possible_match_idx]) {
```

Comment:
Substring search is tricky to do both correctly and efficiently. Use `strstr` (standard) or `memmem` (nonstandard).

If you intend to implement NULs in the `input_str` but not `split_on`, you can use `strstr` and restart the search if a NUL is encountered before end of string. If you intend to implement NULs in both, you may need to implement the Boyer-Moore string search algorithm (or similar).

Reply:
how would strstr() be helpful if we are allowing binary splits?

Reply:
I think this whole thing get's simpler if we don't support NULL and I don't have a use case for binary data. For binary data, I'd typically not be reading and then splitting. It would usually have some sort of framing.... I'd be content to just make this a string-only based API.

Reply:
I'm deleting the string search. Basically I refuse to put unsafe code in our common-lib, and strstr is fundamentally unsafe. I don't want to waste Amazon's money by spending hours reemplementing boyer-moore. And now that I think about it, this function wouldn't be that helpful for http anyways (seeing as how I don't have a way of knowing where the body would begin ahead of time anyways.

I'm deleting that function, will add it when we need it.